add kind e2e setup

sallyom opened the pull request:
This PR adds a kind setup to deploy the charts. This will be used to test the deployment of the stack along with the scaffold jobs. It can also test that tuf.localhost, rekor.localhost are accessible. It will not be suitable for testing the signing & verifying functionality. 

TODO: Add the tests for pod health, job completion, connectivity

/cc @Gregory-Pereira
/cc @sabre1041

lance commented:
@sallyom thanks for doing this! Can you look into this linting error?

```
 ✖︎ trusted-artifact-signer => (version: "0.1.1", path: "charts/trusted-artifact-signer") > chart version not ok. Needs a version bump! 
```

Gregory-Pereira commented:
Encountering other errors as well:

1.  running the kind setup script: `./kind/kind-up-testing.sh`
error logs:
```bash
error: error reading ./kind/test-keys-cert/file_ca_key.pem: no such file or directory
error: no objects passed to apply
error: error reading ./kind/test-keys-cert/rekor_key.pem: no such file or directory
error: no objects passed to apply
```
2. Actual chart installation:
```bash
$ OPENSHIFT_APPS_SUBDOMAIN=localhost envsubst <  ./examples/values-kind-sigstore.yaml | helm upgrade -i trusted-artifact-signer --debug ./charts/trusted-artifact-signer -n sigstore --create-namespace --values -

Error: unable to build kubernetes objects from release manifest: [resource mapping not found for name: "fulcio-system" namespace: "fulcio-system" from "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1"
ensure CRDs are installed first, resource mapping not found for name: "rekor-system" namespace: "rekor-system" from "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1"
ensure CRDs are installed first]
[resource mapping not found for name: "fulcio-system" namespace: "fulcio-system" from "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1"
ensure CRDs are installed first, resource mapping not found for name: "rekor-system" namespace: "rekor-system" from "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1"
ensure CRDs are installed first]
unable to build kubernetes objects from release manifest
```. At the end of this last error message it produced a stacktrace which I didnt find relevant. Looking into this more.

Gregory-Pereira commented on kind/kind-up-test.sh:
```diff
@@ -0,0 +1,46 @@
+
+sudo kind get kubeconfig > /tmp/config
+sudo chown $USER:$USER /tmp/config
+mv /tmp/config ~/.kube/config
```
Should we do some conditional checking here if they have an existing `kubeconfig`, and if so add an entry for this kind cluster, or is this not an issue? Upon trying to debug the script, I realized that my existing quite extensive `kubeconfig` got wiped.

sallyom replied:
oh shoot sorry Greg!! Definitely need that

sallyom replied:
I updated the kind setup script to merge kubeconfigs rather than overwrite

Gregory-Pereira commented on kind/kind-up-test.sh:
```diff
@@ -0,0 +1,46 @@
+
+oc create ns fulcio-system
+oc create ns rekor-system
+oc -n fulcio-system create secret generic fulcio-secret-rh --from-file=private=./kind/test-keys-cert/file_ca_key.pem --from-file=public=./kind/test-keys-cert/file_ca_pub.pem --from-file=cert=./kind/test-keys-cert/fulcio-root.pem  --from-literal=password=secure --dry-run=client -o yaml | oc apply -f-
```
How were the files in this directory (`./kind/test-keys-cert/`) and the directory itself generated? I dont see them provided in the repo and I am not sure how to reproduce them.

sallyom replied:
I must have added a gitignore for those - I think we can push them & use/re-use them since they are just for testing (unless we can't/shouldn't) ... I'll push them just a sec

sallyom replied:
also I added a rh-gitleaks-precommit so we don't get bugged about leaking sensitive data

sallyom commented:
@lance  @Gregory-Pereira 

This should be working now. 

@JasonPowr I added a commit to update the cosign deployment & I added a cosign SA for the rolebinding - I was seeing a configCreate error because of the runAsNonRoot without a runAsUser :woman_shrugging:  & also I modified docs and env to be more generic/less dependent on Keycloak - ptal!

sallyom commented on .gitleaks.toml:
```diff
@@ -0,0 +1,8 @@
+  # Ignore based on any subset of the file path
+  paths = [
+    # Ignore all example certs
+    '''\/testing-only-cert-key\/*\.pem$''',
```
@lance this will prevent warnings from rh that we've leaked sensitive data due to the testing-only cert/keys added for the kind test setup

JasonPowr commented on charts/trusted-artifact-signer/templates/cosign-deployment.yaml:
```diff
@@ -12,27 +12,29 @@ spec:
           value: "https://tuf.$(OPENSHIFT_APPS_SUBDOMAIN)"
-        - name: KEYCLOAK_OIDC_ISSUER
-          value: "$KEYCLOAK_URL/auth/realms/$(KEYCLOAK_REALM)"
         securityContext:
```
@sallyom when testing the cosign changes, I noticed that although not creating an error, the cosign deployment was not properly being deployed you can find this msg in the yaml file of the cosign deployment:

```
      reason: FailedCreate
      message: >-
        pods "cosign-96df6bd59-" is forbidden: unable to validate against any
        security context constraint:
        [pod.metadata.annotations[container.seccomp.security.alpha.kubernetes.io/cosign]:
        Forbidden: seccomp may not be set, provider restricted-v2:
        .containers[0].runAsUser: Invalid value: 65533: must be in the ranges:
        [1001350000, 1001359999], provider "restricted": Forbidden: not usable
        by user or serviceaccount, provider "nonroot-v2": Forbidden: not usable
        by user or serviceaccount, provider "nonroot": Forbidden: not usable by
        user or serviceaccount, provider "pcap-dedicated-admins": Forbidden: not
        usable by user or serviceaccount, provider "hostmount-anyuid":
        Forbidden: not usable by user or serviceaccount, provider
        "hostnetwork-v2": Forbidden: not usable by user or serviceaccount,
        provider "hostnetwork": Forbidden: not usable by user or serviceaccount,
        provider "hostaccess": Forbidden: not usable by user or serviceaccount,
        provider "node-exporter": Forbidden: not usable by user or
        serviceaccount, provider "privileged": Forbidden: not usable by user or
        serviceaccount]
```

I believe this is because the securityContext is not being set properly as its inside the containers array, and should be in the spec object, also somethings cant be set here so they need to be removed.

```
    spec:
      containers:
      securityContext:
        runAsNonRoot: true
        runAsUser: 65533
```

sallyom replied:
hm... ok i'll update that, thank you!

JasonPowr replied:
yea I thought it was weird too alright

JasonPowr replied:
Really sorry @sallyom , looking at the way I made my comment it looks confusing, 

```
      securityContext:
        runAsNonRoot: true
        runAsUser: 65533
```

These are the fields that should be kept in, again apologies

sallyom replied:
oh! :eyes: so there's 
```yaml
spec:
  securityContext:
    outside-container-spec
   containers:
     securityContext:
       inside-container-spec
```
so which lines should be where? 
I believe all of this info should be somewhere:
```
allowPrivilegeEscalation: false
runAsNonRoot: true
capabilities:
  drop:
  - ALL
runAsNonRoot: true
runAsUser: 65533
```
so which goes where? i'm confused - as/is things are working in Kind - but probably you're seeing fails in OpenShift, right?

sallyom commented:
@JasonPowr I updated the cosign deployment _and_ added a `helm test` check it out!! Look at the output, it's beautiful:
```
$ helm test -n sigstore trusted-artifact-signer
NAME: trusted-artifact-signer
LAST DEPLOYED: Tue Oct  3 09:56:37 2023
NAMESPACE: sigstore
STATUS: deployed
REVISION: 1
TEST SUITE:     trusted-artifact-signer-test-connection
Last Started:   Tue Oct  3 10:00:34 2023
Last Completed: Tue Oct  3 10:00:37 2023
Phase:          Succeeded
```

sallyom commented on .github/workflows/test.yaml:
```diff
@@ -57,8 +57,21 @@ jobs:
+      shell: bash
+
+      - name: Run chart-testing (install)
+        run: ct install --helm-extra-args "--wait --wait-for-jobs" --config ct-install.yaml
```
I have no idea if this workflow works - we'll iterate on it

sallyom commented on .github/workflows/test.yaml:
```diff
@@ -57,8 +57,21 @@ jobs:
+#     shell: bash
+
+#     - name: Run chart-testing (install)
+#        run: ct install --helm-extra-args "--wait --wait-for-jobs" --config ct-install.yaml
```
commented out for now but don't want to delete it, it will be un-commented in a follow-up

lance commented:
/lgtm
